Refactor ColorSprayBehaviour using BFS implementation

omergunr100 opened the pull request:
## What
Spray paint only paints pipes/cables/paintables of the same color as the original block when painting multiple blocks.
Spray paint will stop painting when 2 adjacent pipes/cables are disconnected instead of ignoring connections. (works on the p2p case too)

## Details
When sneaking, instead of checking blocks in a line from the player switched to breadth first search implementation, blocks are now collected then painted. (potential memory overhead of ~50KB in worst case using maximum config value for block limit)

## Additionally
Resolves #2518
Fixes #2570

omergunr100 committed:
better painting of ae cables and gt pipes

omergunr100 committed:
forgot mm check

omergunr100 committed:
and another one

tomprince commented on src/main/java/com/gregtechceu/gtceu/common/item/ColorSprayBehaviour.java:
```diff
@@ -329,28 +342,40 @@ private boolean tryPaintSpecialBlock(Player player, Level world, BlockPos pos, B
+            if (be instanceof CableBusBlockEntity cable && first instanceof CableBusBlockEntity og) {
+                var side = getPaintDirection(player).getOpposite();
+                if (!first.getBlockPos().equals(be.getBlockPos()) &&
+                        (cable.getPart(side) != null || cable.getCableConnectionType(side) == AECableType.NONE)) {
```
If I am reading this right, for each block, it checks if near side has a cable part, and if so, stops before painting that block. I *think* you also want to check is the *previous* block has a cable part on the far side.

omergunr100 replied:
for both ae2 cables and gt pipes it's almost the same thing iirc, if the previous cable has a part on the far side the near side of the current one will either be disconnected or have another part.

tomprince replied:
With no knowledge of how AE represents things internally, is that true of ME P2P tunnel? Certainly, from a logical perspective, the cable opposite the tunnel is connected to the tunnel. I don't know what that means in terms of `getCableConnectionType` though.

omergunr100 replied:
you're right, p2p is an edge case, I'll see how it can be handled

omergunr100 replied:
the right way to do it would probably be a refactor of the behaviour to use some custom collector logic for each block paintable block type, that would also help if we want to paint the connected network instead of just a straight line, I'll give it a shot later

omergunr100 replied:
fixed

omergunr100 committed:
conditional bfs stuff

omergunr100 committed:
behaviour refactor

omergunr100 committed:
removed logger

omergunr100 committed:
Merge branch '1.20.1' into 1.20.1-paint

omergunr100 commented:
hold on, I forgot the case for meta machines

omergunr100 committed:
meta machine special case.
moved useOn to first so that ui machines will be paintable.
cleanup and formatting.

omergunr100 committed:
missed the ! symbol

omergunr100 committed:
remove leftover create support failed feature

omergunr100 committed:
leftovers from seat

omergunr100 committed:
fix shulker box losing its inventory when painted

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/item/ColorSprayBehaviour.java:
```diff
@@ -513,4 +496,49 @@ public static void setUsesLeft(ItemStack itemStack, int usesLeft) {
+        return parent instanceof IPaintable pp && child instanceof IPaintable pc && paintablePredicate.test(pp, pc);
+    };
+
+    private static final TriPredicate<CableBusBlockEntity, CableBusBlockEntity, Direction> ae2CablePredicate = (parent,
+                                                                                                                child,
+                                                                                                                direction) -> {
+        if (parent == null) return true;
+        var childDirection = direction.getOpposite();
+        if (parent.getPart(direction) != null || parent.getCableConnectionType(direction) == AECableType.NONE ||
+                child.getPart(childDirection) != null ||
+                child.getCableConnectionType(childDirection) == AECableType.NONE ||
+                parent.getColor() != child.getColor()) {
+            return false;
+        }
+        return true;
+    };
```
wrap call in nested class to not get class load exception

omergunr100 committed:
wrapped

tomprince commented:
A couple of comments after testing this:
- I was a bit surprised that the new behavior paints multiple types of AE2 cables of the same color. But I think this is mostly because I was familiar with the previous implementation, and that the new behavior is sensible. I haven't used the spray can for any other kinds of blocks, so I don't know if the more generous (in terms of block identity check) makes sense for the other special types.
- I was wondering about how painting should interact with cables of the paint color and uncolored:
  - Should blocks that are already the paint color stop the search, or should it continue looking for blocks of the initial painted color further along.
  - If you try painting a cable that is already the painted color, should it trying searching for unconnected uncolored cables and paint those?
  
  Both of those last two are perhaps a bit too much magic though.

tomprince commented:
I will say that this does make the spray can operate on AE2 cables like I originally expected to. :tada:

omergunr100 commented:
> * I was wondering about how painting should interact with cables of the paint color and uncolored:
>   
>   * Should blocks that are already the paint color stop the search, or should it continue looking for blocks of the initial painted color further along.
>   * If you try painting a cable that is already the painted color, should it trying searching for unconnected uncolored cables and paint those?
>   
>   Both of those last two are perhaps a bit too much magic though.

I considered these options and even implemented the first one initially but it was just too weird feeling.
The uncoloured search one isn't consistent with the expected behaviour imo, to a non-technical player "uncoloured" is just another colour.
